algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py

```diff
@@ -0,0 +1,348 @@
+    pass
+
+  def loss_fn(self,
+              label_batch: spec.Tensor,
+              logits_batch: spec.Tensor,
+              mask_batch: Optional[spec.Tensor] = None) -> spec.Tensor:
+
+    per_example_losses = metrics.per_example_sigmoid_binary_cross_entropy(
+            logits = logits_batch, targets=label_batch)
+
+    if mask_batch is not None:
+        weighted_losses = per_example_losses * mask_batch
+        normalization = mask_batch.sum()
+
+    else:
+        weighted_losses = per_example_losses
+    normalization = label_batch.shape[0]
+
+    return torch.sum(weighted_losses, dim=-1) / normalization
+
+    #return F.binary_cross_entropy_with_logits(
+    #    logits_batch.reshape(-1, 1),
+    #    label_batch.reshape(-1, 1),
+    #    reduction='none')
+
+  def _eval_metric(self, logits: spec.Tensor,
+                   targets: spec.Tensor) -> Dict[str, int]:
+
+    loss = self.loss_fn(logits, targets).sum()
+    return {'loss': loss}
+
+  def _eval_model_on_split(self,
+                           split: str,
+                           num_examples: int,
+                           global_batch_size: int,
+                           params: spec.ParameterContainer,
+                           model_state: spec.ModelAuxiliaryState,
+                           rng: spec.RandomState,
+                           data_dir: str) -> Dict[str, float]:
+    """Evaluate the model on a given dataset split, return final scalars."""
+    data_rng, model_rng = prng.split(rng, 2)
+
+    if split not in self._eval_iters:
+      self._eval_iters[split] = self.build_input_queue(
+          data_rng, split, data_dir, global_batch_size=global_batch_size)
+
+    total_metrics = {
+        'loss': torch.tensor(0., device=DEVICE),
+    }
+
+    num_batches = int(math.ceil(num_examples / global_batch_size))
+
+    logits_list = []
+    targets_list = []
+
+    for _ in range(num_batches):
+      batch = next(self._eval_iters[split])
+      logits, _ = self.model_fn(params, batch, model_state,
+                                spec.ForwardPassMode.EVAL, model_rng, False)
+
+      batch_metrics = self._eval_metric(logits, batch['targets'])
+      total_metrics = {
+          k: v + batch_metrics[k] for k, v in total_metrics.items()
+      }
+
+      # all batch eval
+      output_receive_buffer = torch.empty((global_batch_size // N_GPUS),
+                                          device=DEVICE)
+      reciever = [output_receive_buffer] * N_GPUS
+      torch.distributed.all_gather(reciever, logits)
+      logits_list.append(output_receive_buffer.float())
+      targets_list.append(batch['targets'])
+    compute_auc = metrics.roc_auc_score(
+        torch.sigmoid(torch.cat(logits_list)), torch.cat(targets_list))
+    if USE_PYTORCH_DDP:
+      for metric in total_metrics.values():
+        dist.all_reduce(metric)
+    computed_val = {
+        k: float(v.item() / num_examples) for k, v in total_metrics.items()
```

Comment:
This is likely not exact, since the last eval batch is dropped, so not all `num_examples` samples are used. But doesn't explain the lack of training performance.